sdks/python/apache_beam/runners/runner.py

```diff
@@ -133,6 +133,26 @@ def visit_transform(self, transform_node):
 
   return GroupByKeyInputVisitor()
 
+def flatten_input_visitor():
+  # Imported here to avoid circular dependencies.
+  from apache_beam.pipeline import PipelineVisitor
+
+  class FlattenInputVisitor(PipelineVisitor):
+    """A visitor that replaces the element type for input ``PCollections``s of a
+     ``Flatten`` transform with that of the output ``PCollection``.
+    """
+
+    def visit_transform(self, transform_node):
+      # Imported here to avoid circular dependencies.
+      # pylint: disable=wrong-import-order, wrong-import-position
+      from apache_beam import Flatten
+      from apache_beam import typehints
+      if (isinstance(transform_node.transform, Flatten)):
+        output_pcoll = transform_node.outputs[None]
+        for input_pcoll in transform_node.inputs:
+          input_pcoll.element_type = output_pcoll.element_type
+
+  return FlattenInputVisitor()
 
 class PipelineRunner(object):
```

Comment:
For the `GroupByKey` it will be useful for other non-python runners too, but for `Flatten` yes I would like it to be in the Dataflow runner. Either way this would be much easier to do once we move to the new runner API. Added a TODO.